### ISSUE TITLE
Extending the list of properties for schema-org:Article 

### ISSUE THREAD

**Issue by cyberandy** [insideout10/wordlift-plugin#371]
As we fixed #262 we've added to all articles the schema-org:Article attribute - this way G+ doesn't mess up entities with articles. As we've added this new attribute to blog post it would be recommended to add to the list of properties for an article:
1. author (we have it)
2. datePublished (we have it) 
3. headline (we have it)
4. image (we have it)
5. publisher (this can be added in the WordLift settings)  
6. dateModified (we have it) 
7. mainEntityOfPage (this corresponds to the entity the user chooses as _Topic_ under the tab "Article metadata") 

This issue shall probably go right after #310 (the switch to JSON-LD) - here is a valid example  http://jsonld.com/article/. We need to test carefully this one with special regards to compatibility with SEO Ultimate, SEO Yoast, All in One SEO Pack e AMP.

**Comment by ziodave** [comment]
We need to clarify the following:

* _author_ it's expecting an entity, should we explicit the `@id` (e.g. http://data.wordlift.it/testwordlift/user/fady) ?
* being the post title the _name_, is the excerpt the _headline_?
* when an _image_ is not set for the post, even though it's required by Google, we wouldn't publish it until an editor selects an image for the post
* _publisher_ is an _Organization_ or a _Person_ so it should be a link to an entity defined in the vocabulary

**Comment by ziodave** [comment]
Some useful reference for Google+: https://developers.google.com/+/web/snippet/article-rendering

**Comment by cyberandy** [comment]
> _author_ it's expecting an entity, should we explicit the @id (e.g. http://data.wordlift.it/testwordlift/user/fady) ?

I believe we shall always deal with entities - so IMO yes we shall use the @id 

> being the post title the name, is the excerpt the headline?


I would keep the title as both _name_ and _headline_ as the Guardian (they use `schema:description` for the lead and this could be the excerpt of the article in our case). There are different practices here (i.e. BBC uses only `schema:headline` for article title and `schema:description` for the article lead)

> when an image is not set for the post, even though it's required by Google

Yes of course this totally makes sense

> _publisher_ is an Organization or a Person so it should be a link to an entity defined in the vocabulary

Yes either an Author or an Organization already created in the vocabulary can be used as _publisher_ (@wanbinkimoon remember that for the new Settings page)

**Comment by novedavi** [comment]
Example from camminandocon of the issue:
![screen shot 2016-11-08 at 10 42 56](https://cloud.githubusercontent.com/assets/18528457/20096072/867ec7ac-a5a8-11e6-8216-2078d5f6c7ec.png)

**Comment by cyberandy** [comment]
In this context it might also good to evaluate how to add - in the Article markup - a link to metadata we published as lod (i.e. http://data.wordlift.it/wl05373330/post/how_to_see_through_the_algorithms_in_your_life). I could not find though yet a proper example on how to implement this in schema.org.

**Comment by ziodave** [comment]
We need to test here if we can migrate schema:Article as JSON-LD instead of microdata.

**Comment by ziodave** [comment]
* for _author_ we try to regex on the _rel=author_ & co. If not found we can output hidden microdata such as ```<span itemprop="author" itemscope="" itemtype="http://schema.org/Person"><meta itemprop="name" content="David R" /></span>```
* for _datePublished_, ```<time itemprop="startDate" datetime="2011-05-08T19:30">```
* for _headline_ we'll use the _title_ as suggested by different web sites, http://www.bloggingspell.com/add-schema-org-markup-wordpress/, http://keithdevon.com/schema-on-your-wordpress-website/, http://orbitingweb.com/blog/schema-org-microdata-guide/
* for the _image_ we'll look into the page for the exact URL of the featured image and, if not found, for any variations of it, e.g. _base-url_-123x123.ext and add the `itemprop="image"`
* for the _publisher_ we'll add the publisher data using spans and metas. We avoid putting empty spaces/new lines between spans and adjacent markup to avoid styling issues.

**Comment by ziodave** [comment]
I'd like to make some compatibility tests here with Yoast SEO, this page for example has 2 itemscopes on the article tag:
http://www.teodorapetkova.com/poiesis-of-relationships/wordlift-linked-data-texts/

I also would like to get a list of the most used WP's theme to check for compatibility of our microdata.

With an eye to #451 and G+ only supporting microdata, we need to ensure that we don't mess the html up.

**Comment by ziodave** [comment]
For the http://www.teodorapetkova.com web site we need to consider that Genesis is also adding the `itemscope itemtype` properties.

We also need to understand if we should be more specific about posts (calling them `BlogPosting`s), although this is not the case for every web site.

**Comment by ziodave** [comment]
We need also to take care about nested articles.

**Comment by ziodave** [comment]
I've been reviewing this issue with @markkap extensively. Using microdata can create many issues as we cannot foresee what each theme is doing, prove is that web sites using Genesis (for example) already output some kind of microdata on the article tag.

Other complexities may arise when there is more than one article tag, or even nested tags, it would make complicated to target the right one.

Adding properties would be another challenge as microdata would require to wrap the visible data when it is available creating another series of potential failures and errors with themes and styles.

So why are we using microdata? We aren't anymore, we were using it to solve G+ issues, #262, when we were used to output the entities data embedded as microdata in page. But since we transitioned to JSON-LD, #310.

Therefore we're changing the target of this issue to output schema:Article/BlogPosting data as JSON-LD.

**Comment by ziodave** [comment]
Tests failing, need to fix this:

```
Fatal error: Call to protected method Wordlift_Abstract_Post_To_Jsonld_Converter::relative_to_context() from context '' in /home/travis/build/insideout10/wordlift-plugin/src/includes/class-wordlift-entity-post-to-jsonld-converter.php on line 105
```

**Comment by cyberandy** [comment]
As I'm running some tests on 3.10 I noticed the following while analysing [this article](http://test.wordlift.it/paterson-un-film-di-jim-jarmusch/): 

![image](https://cloud.githubusercontent.com/assets/837037/22585494/615cb3dc-e9f8-11e6-8a10-39d4a22a3920.png)

[Here](https://search.google.com/structured-data/testing-tool/u/0/?url=http://test.wordlift.it/paterson-un-film-di-jim-jarmusch/#url=http%3A%2F%2Ftest.wordlift.it%2Fpaterson-un-film-di-jim-jarmusch%2F) is the details from the Google Structured Data testing tool (1 error and 1 warning under BlogPosting - Article is all right)

![image](https://cloud.githubusercontent.com/assets/837037/22585603/de1bc110-e9f8-11e6-872a-c65b5b47fc99.png)

**Comment by cyberandy** [comment]
I have found another anomaly over [here](http://test.wordlift.it/vocabulary/jim-jarmusch/) this is an entity `type=person` but the only the Google Structured Data testing tool only sees the creative works associated with the entity. See [here](https://search.google.com/structured-data/testing-tool/u/0/?url=http://test.wordlift.it/vocabulary/jim-jarmusch/#url=http%3A%2F%2Ftest.wordlift.it%2Fvocabulary%2Fjim-jarmusch%2F) for more details.

**Comment by ziodave** [comment]
We don't use _BlogPosting_, as per our decision [here](https://github.com/insideout10/wordlift-plugin/issues/371#issuecomment-273789358). The _BlogPosting_ with the missing _publisher_ is added by the *Schema App Structured Data* plugin which I now disabled on the test web site:

![image](https://cloud.githubusercontent.com/assets/11438/22599844/87330b70-ea38-11e6-8b26-61c2478e0ff8.png)

As for the second issue it is more subtle, since we're actually outputting the right data:

![image](https://cloud.githubusercontent.com/assets/11438/22600017/1c78de6c-ea39-11e6-93d1-99e4c36889e1.png)

**Comment by ziodave** [comment]
The 2nd also is correct: in the entity's content the "Paterson" _CreativeWork_ is mentioned. "Paterson" has "[Jim Jarmush](http://data.wordlift.it/testwordlift/entity/jim_jarmusch)" as _author_.

Therefore Google Structured Data Tool displays the person linked in the _author_ property of "Paterson":

![image](https://cloud.githubusercontent.com/assets/11438/22600474/dfce60ca-ea3a-11e6-9128-67f3b43a4417.png)

I'll mark the issue as _resolved_.

**Comment by ziodave** [comment]
We can always decide that the JSON-LD for entities doesn't include entities in the post content, but only those in properties.

**Comment by cyberandy** [comment]
ok perfect - apologies for the confusion with Schema App (I should have
think of it)

**Comment by ziodave** [comment]
I was checking if Person has a transitive property for CreativeWork/author but couldn't find any.

**Comment by inetbiz** [comment]
Just added plugin to WP Genesis framework site. I get errors in article schema for publisher. See https://goo.gl/9Gfy7f

**Comment by ziodave** [comment]
Hello @inetbiz, I've created a new issue for this:
https://github.com/insideout10/wordlift-plugin/issues/471

**Comment by FitBoyAwesome** [comment]
Solid article my man! I wish I would have discovered sooner for my site [canadaprimemarketing.com](https://www.canadaprimemarketing.com). I already put in all the groundwork and discovered all of this and implementing it about 2 weeks ago. But totally on point none the less!
